### ISSUE TITLE
Move to HTML5 Player

### ISSUE THREAD

**Issue by EugeneKay** [Libresonic/libresonic#45]
Moving to a HTML5 player is desired. Patches welcome.

**Comment by jolny** [comment]
Could http://jplayer.org/ be an alternative? It's under a MIT license.

**Comment by csobsidian** [comment]
Native HTML5 audio? or are we talking video?

**Comment by apastuszak** [comment]
I believe Ampache uses jplayer, but I could be wrong.

**Comment by jolny** [comment]
@csobsidian it appears it supports both :)

**Comment by csobsidian** [comment]
a list of html5 video/audio players and their FOSS licenses:

http://html5video.org/wiki/HTML5_Video_Player_Comparison

**Comment by csobsidian** [comment]
It's more work but we could write our own interface and use the [Web Audio API](https://www.w3.org/TR/webaudio/), which is supported by [all current major browsers](https://en.wikipedia.org/wiki/HTML5_Audio#Web_Audio_API_and_MediaStream_Processing_API). This option would allow us to do bumpless playback by pre-buffering the next audio file, software EQ in browser, and even audio aware themes.

**Comment by RomanHargrave** [comment]
@csobsidian's idea is my favorite

**Comment by RomanHargrave** [comment]
To follow up, I just switched subsonic (the kang libresonic came from) to libresonic, and found to my (disgust) that this using a flash player. Good lord.

**Comment by apastuszak** [comment]
The kang has always used as flash player, as has Subsonic prior to version
6.

**Comment by RomanHargrave** [comment]
@apastuszak I was able to use it in a browser without flash support. It appeared to be using JWPlayer, which will use HTML5 audio if the browser supports it.

**Comment by apastuszak** [comment]
You were probably running 6.0beta1 of kang.  I believe the kang version had
to roll back to 5.3 for the Libresonic release, since that was the last
version that was GPL licensed.

**Comment by RomanHargrave** [comment]
On Thu, 12 May 2016 11:50:28 -0700
apastuszak <notifications@github.com> wrote:

> You were probably running 6.0beta1 of kang.  I believe the kang version had
> to roll back to 5.3 for the Libresonic release, since that was the last
> version that was GPL licensed.
> 
> On Thu, May 12, 2016 at 2:32 PM, Roman Hargrave <notifications@github.com>
> wrote:
> 
> > @apastuszak <https://github.com/apastuszak> I was able to use it in a
> > browser without flash support. It appeared to be using JWPlayer, which will
> > use HTML5 audio if the browser supports it.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/Libresonic/libresonic/issues/45#issuecomment-218845858>
> >  
> 
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/Libresonic/libresonic/issues/45#issuecomment-218850816

Ah.

Well, knowing about the existence of the SVN copy helps.

I'll see about opening a PR to fix this if that feature is in a mergeable state

-- 
Roman Hargrave <roman@hargrave.info>
http://hargrave.info
 _________________________________________
/ The only thing worse than X Windows: (X \
\ Windows) - X                            /
 -----------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

**Comment by woodruffw** [comment]
Looks like someone hacked a JS version of JWPlayer into Subsonic here:
http://forum.subsonic.org/forum/viewtopic.php?f=6&t=16239

That solution probably isn't preferable (at a glimpse, licensing keys + proprietary player when equally good/better FOSS alternatives exist), but it's a starting point.

**Comment by sevensins** [comment]
I am trying to patch in Amplitude [https://github.com/521dimensions/amplitudejs] with FontAwesome icon for the player controls.
Will push in a patch once done.

**Comment by muff1nman** [comment]
It doesn't seem from a first glance that amplitude J's supports video. Ideally we could use one external framework for both video and audio

**Comment by sevensins** [comment]
@muff1nman : Do we really want to run video using Libresonic ?? And yes amplitude doesn't support video.

**Comment by muff1nman** [comment]
Yes, it's something that is currently supported and I am not aware of any plans to move away from supporting it. I think it's awesome you're excited to integrate amplitude Js but I think it might be best if we pause for a bit to ensure we select a good strategy that will support libresonic in the long term before we start doing any large PRs.

**Comment by biconou** [comment]
I want to keep video.

**Comment by sevensins** [comment]
@muff1nman : If there is already support for video (sorry never tested that, since I have always used to stream audio {{ sonic == sound }}), I guess it's better to have continuation for video support.
But just my opinion, if video is really needed there is Kodi / Plex  ....
With the name Libresonic, I feel that audio is what should be emphasized upon. Then again this is what I feel and my personal opinion.

**Comment by muff1nman** [comment]
Audio is emphasized on. However, libresonic also does a great job with video and is something we should keep.

**Comment by raulvasquez** [comment]
I used to have perisonic running as my default page http://robinbakker.nl/ I haven't looked at the licenses, but maybe it could be something to build off of?

**Comment by muff1nman** [comment]
https://github.com/johndyer/mediaelement seems promising

**Comment by biconou** [comment]
I think that the support of lossless formats (FLAC) is important.

**Comment by muff1nman** [comment]
That's the whole point of having a transcoder. As long as it supports a few decent formats, we can transcode on the fly to those formats.

**Comment by Eusmilis** [comment]
I am currently experimenting with an <audio> tag. It seems to be working, but the UI is not consistent between browsers.

**Comment by RomanHargrave** [comment]
Out of boredom, and curiosity as to why opus isn't working, I think I may look in to replacing jwplayer with jplayer.

**Comment by izderadicka** [comment]
Opus support is also important for me.   Currently it works for me fine in Chromium, where HTML5 player is used,  but problem is in Firefox where flash player is enforced.  I think key problem is not in jwPlayer per se, but rather in fact that version of jwPlayer used is really ancient, not reflecting current capabilities of modern browser. 

I think HTML5 player is the only one aspect of whole Web interface issue - which is kind of outdated, based on technologies older then 5 years (jquery 1.7, DWR, JSPs pages messing everything together  - markup, JS, server code).   So rather then thinking about partial upgrades to fairly outdated WebUI, why not to think about completely new face using recent technologies - like React, AngularJS, Aurelia etc. Maybe split interface into two parts - admin part, which might stick to older type interface, be more conservative and player part (dependent only on REST API), which can leverage latest browser technologies and possibly offer several player alternatives?

**Comment by muff1nman** [comment]
@izderadicka a full upgrade is a ton of work. We don't have the man power to do it. As discussed on Reddit, I am leaning to incremental upgrades to the existing ui.

**Comment by RomanHargrave** [comment]
@muff1nman I've seen single people do some impressive interface design, given enough motivation.

What's to say someone won't open a PR with an objectively superior web UI?

**Comment by muff1nman** [comment]
@RomanHargrave I'm not saying it won't happen, or that I wouldn't be happy to see a PR, rather I'm saying that our current contributor efforts might be better spent distributed amongst a host of things compared a full ui upgrade. However let's not hijack this thread further and take more long term discussion elsewhere.

**Comment by DArtagan** [comment]
Here are the ones I came up with that had compatible licenses and features.

I think these two are especially worth considering:
* [Video.js](https://github.com/videojs/video.js)
* [Plyr](https://github.com/selz/plyr)

As alternatives these two deserve mentioning as passable:
* [afterglow](https://github.com/moay/afterglow)
* [MediaElement.js](https://github.com/mediaelement/mediaelement)
